### PR TITLE
Replace deprecated sklearn plotting functions with plotting class methods

### LINKED ISSUE #5569: Replace scikit-learn plotting fuctions with plotting class methods
Plotting functions in scikit-learn (e.g. `plot_confusion_matrix`) are deprecated in 1.0 and will be removed in 1.2. They should be replaced with plotting class methods (e.g. `ConfusionMatrixDisplay.from_predictions`).

https://github.com/mlflow/mlflow/blob/1ddb2c9b5ace0fa605195a4b14c595e274a8c384/mlflow/sklearn/utils.py#L313

For example, the code above can be fixed as follows:

```python
if Version(sklearn.__version__) >= Version("1.0"):
    return ConfusionMatrixDisplay.from_predictions((*args, **kwargs) )
else:
    return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
```

## Deprecation warnings

https://github.com/mlflow/mlflow/runs/5766838748?check_suite_focus=true#step:12:188

```
  /opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function plot_confusion_matrix is deprecated; Function `plot_confusion_matrix` is deprecated in 1.0 and will be removed in 1.2. Use one of the class methods: ConfusionMatrixDisplay.from_predictions or ConfusionMatrixDisplay.from_estimator.
    warnings.warn(msg, category=FutureWarning)

tests/sklearn/test_sklearn_autolog.py::test_eval_and_log_metrics_for_binary_classifier
tests/sklearn/test_sklearn_autolog.py::test_eval_and_log_metrics_matches_training_metrics
  /opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function plot_roc_curve is deprecated; Function :func:`plot_roc_curve` is deprecated in 1.0 and will be removed in 1.2. Use one of the class methods: :meth:`sklearn.metric.RocCurveDisplay.from_predictions` or :meth:`sklearn.metric.RocCurveDisplay.from_estimator`.
    warnings.warn(msg, category=FutureWarning)

tests/sklearn/test_sklearn_autolog.py::test_eval_and_log_metrics_for_binary_classifier
tests/sklearn/test_sklearn_autolog.py::test_eval_and_log_metrics_matches_training_metrics
  /opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function plot_precision_recall_curve is deprecated; Function `plot_precision_recall_curve` is deprecated in 1.0 and will be removed in 1.2. Use one of the class methods: PrecisionRecallDisplay.from_predictions or PrecisionRecallDisplay.from_estimator.
    warnings.warn(msg, category=FutureWarning)
```

### PR DESCRIPTION
**Pull request by shaikmoeed**
…ions

## What changes are proposed in this pull request?

Fix #5569 

Replaced 
https://github.com/mlflow/mlflow/blob/1ddb2c9b5ace0fa605195a4b14c595e274a8c384/mlflow/sklearn/utils.py#L313

with 

```
if Version(sklearn.__version__) >= Version("1.0"):
    return ConfusionMatrixDisplay.from_predictions((*args, **kwargs) )
else:
    return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
```

## How is this patch tested?

(Details)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/utils.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import logging
 from numbers import Number
 import numpy as np
+from sklearn.metrics import ConfusionMatrixDisplay
```
Comment (harupy) [inline mlflow/sklearn/utils.py:7]: Can we import this in `_get_classifier_artifacts`?

Comment (harupy) [reply mlflow/sklearn/utils.py:7]: If we import `ConfusionMatrixDisplay` here, users who don't need to use `mlflow.sklearn` need to install `sklearn`.

File: mlflow/sklearn/utils.py [hunk 2]
```diff
@@ -310,7 +311,10 @@ def plot_confusion_matrix(*args, **kwargs):
                 "axes.labelsize": 10.0,
             }
         ):
-            return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            if Version(sklearn.__version__) >= Version("1.0"):
+                return ConfusionMatrixDisplay.from_predictions(*args, **kwargs)
+            else:
+                return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:317]: Thanks for the change! I think there are a few more plotting functions. Can you also fix them?

File: mlflow/sklearn/utils.py [hunk 3]
```diff
@@ -310,7 +311,10 @@ def plot_confusion_matrix(*args, **kwargs):
                 "axes.labelsize": 10.0,
             }
         ):
-            return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            if Version(sklearn.__version__) >= Version("1.0"):
```
Comment (harupy) [inline mlflow/sklearn/utils.py:314]: It looks like we need to adjust the version here:

https://github.com/mlflow/mlflow/runs/6032083575?check_suite_focus=true#step:12:175

```
E   ImportError: cannot import name 'ConfusionMatrixDisplay' from 'sklearn.metrics' (/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/sklearn/metrics/__init__.py)
```

File: mlflow/sklearn/utils.py (hunk 2 above)
Comment (harupy) [reply mlflow/sklearn/utils.py:317]:
```
2022/04/14 23:14:39 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function plot_roc_curve is deprecated; Function :func:`plot_roc_curve` is deprecated in 1.0 and will be removed in 1.2. Use one of the class methods: :meth:`sklearn.metric.RocCurveDisplay.from_predictions` or :meth:`sklearn.metric.RocCurveDisplay.from_estimator`."
2022/04/14 23:14:39 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function plot_precision_recall_curve is deprecated; Function `plot_precision_recall_curve` is deprecated in 1.0 and will be removed in 1.2. Use one of the class methods: PrecisionRecallDisplay.from_predictions or PrecisionRecallDisplay.from_estimator."
2022/04/14 23:14:41 WARNING mlflow.sklearn.utils: _get_classifier_artifacts.<locals>.plot_confusion_matrix failed. The artifact val_confusion_matrix will not be recorded. Artifact error: from_predictions() got an unexpected keyword argument 'estimator'
```

File: mlflow/sklearn/utils.py (hunk 3 above)
Comment (harupy) [inline mlflow/sklearn/utils.py:314]:
```suggestion
            if Version(sklearn.__version__) >= Version("1.0"):
                from sklearn.metrics import ConfusionMatrixDisplay
```

Can we import `ConfusionMatrixDisplay` here?

File: mlflow/sklearn/utils.py [hunk 4]
```diff
@@ -310,7 +311,10 @@ def plot_confusion_matrix(*args, **kwargs):
...
+                return ConfusionMatrixDisplay.from_predictions(*args, **kwargs)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:315]:
```suggestion
                return ConfusionMatrixDisplay.from_estiamtor(*args, **kwargs)
```

Can we `from_estiamtor` instead?

**Comment by harupy** [comment]
Thanks @shaikmoeed for the PR, left some comments!

File: mlflow/sklearn/utils.py (hunk 1 above)
Comment (harupy) [inline mlflow/sklearn/utils.py:7]:
```suggestion
```

Let's remove this line.

File: mlflow/sklearn/utils.py (hunk 2 above)
Comment (harupy) [reply mlflow/sklearn/utils.py:317]: Or we can focus on `plot_confusion_matrix` in this PR, and then fix the rest of the warnings in a separate PR.

Comment (harupy) [reply mlflow/sklearn/utils.py:317]: @shaikmoeed Or we can fix `plot_confusion_matrix` first and then fix `plot_roc_curve` and `plot_precision_recall_curve` later.

**Comment by harupy** [comment]
@shaikmoeed Thanks for the update! Do you want to fix other plotting functions in this PR?

File: mlflow/sklearn/utils.py (hunk 2 above)
Comment (shaikmoeed) [reply mlflow/sklearn/utils.py]: Fixed one by one warnings for `plot_confusion_matrix`, `plot_roc_curve` and `plot_precision_recall_curve` plotting functions.

File: examples/h2o/random_forest.ipynb [hunk 5]
```diff
@@ -94,12 +94,13 @@
     "import mlflow\n",
     "import mlflow.h2o\n",
     "from h2o.estimators.random_forest import H2ORandomForestEstimator\n",
+    "\n",
```
Comment (harupy) [inline examples/h2o/random_forest.ipynb:97]: Can we revert the unrelated changes?

**Comment by shaikmoeed** [comment]
> @shaikmoeed Thanks for the update! Do you want to fix other plotting functions in this PR?

Yes, I have fixed other plotting functions.

File: examples/h2o/random_forest.ipynb (hunk 5 above)
Comment (shaikmoeed) [reply examples/h2o/random_forest.ipynb:97]: Okay, I will revert those changes. Black autoformatted them.

File: mlflow/sklearn/utils.py [hunk 6]
```diff
@@ -335,7 +340,9 @@ def plot_confusion_matrix(*args, **kwargs):
             [
                 _SklearnArtifact(
                     name=prefix + "roc_curve",
-                    function=sklearn.metrics.plot_roc_curve,
+                    function=sklearn.metric.RocCurveDisplay.from_estimator
```
Comment (harupy) [inline mlflow/sklearn/utils.py:343]:
```suggestion
                    function=sklearn.metrics.RocCurveDisplay.from_estimator
```

File: mlflow/sklearn/utils.py [hunk 7]
```diff
@@ -310,7 +310,12 @@ def plot_confusion_matrix(*args, **kwargs):
                 "axes.labelsize": 10.0,
             }
         ):
-            return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            if Version(sklearn.__version__) >= Version("1.0"):
```
Comment (harupy) [inline mlflow/sklearn/utils.py:313]: Can we put `Version(sklearn.__version__) >= Version("1.0")` in a variable (e.g. `is_plotting_function_deprecated`) and reuse it?

File: mlflow/sklearn/utils.py [hunk 8]
```diff
@@ -328,14 +333,18 @@ def plot_confusion_matrix(*args, **kwargs):
         ),
     ]
 
+    is_plot_function_deprecated = Version(sklearn.__version__) >= Version("1.0")
```
Comment (harupy) [inline mlflow/sklearn/utils.py:336]: Can we move this function at line 297 and use it `plot_confusion_matrix` as well?

File: mlflow/sklearn/utils.py [hunk 9]
```diff
@@ -310,7 +310,12 @@ def plot_confusion_matrix(*args, **kwargs):
                 "axes.labelsize": 10.0,
             }
         ):
-            return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            if Version(sklearn.__version__) >= Version("1.0"):
+                from sklearn.metrics import ConfusionMatrixDisplay
+
+                return ConfusionMatrixDisplay.from_estiamtor(*args, **kwargs)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:316]:
```suggestion
                return sklearn.metrics.ConfusionMatrixDisplay.from_estiamtor(*args, **kwargs)
```

**Comment by harupy** [comment]
@shaikmoeed Thanks for the quick update, can you fix the DCO check failure?

**Comment by shaikmoeed** [comment]
> @shaikmoeed Thanks for the quick update, can you fix the DCO check failure?

Fixed!

**Comment by harupy** [comment]
Thanks! I think all the sklearn tests should pass now :) Let's wait for them to finish, confirm no warnings are raised, and then merge the PR!

File: mlflow/sklearn/utils.py [hunk 10]
```diff
@@ -310,7 +312,11 @@ def plot_confusion_matrix(*args, **kwargs):
                 "axes.labelsize": 10.0,
             }
         ):
-            return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            return (
+                sklearn.metrics.ConfusionMatrixDisplay.from_estiamtor(*args, **kwargs)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:316]:
```suggestion
                sklearn.metrics.ConfusionMatrixDisplay.from_estimator(*args, **kwargs)
```

can we fix a typo?

Comment (harupy) [reply mlflow/sklearn/utils.py]: fixed!

Comment (shaikmoeed) [reply mlflow/sklearn/utils.py]: Thanks! Sorry for the typo error.

**Comment by harupy** [comment]
`Cross version tests / test (sklearn / 1.0.2 / autologging)` failed again:

https://github.com/mlflow/mlflow/runs/6036800236?check_suite_focus=true#step:12:866

```
2022/04/15 10:19:59 WARNING mlflow.sklearn.utils: _get_classifier_artifacts.<locals>.plot_confusion_matrix failed. The artifact val_confusion_matrix will not be recorded. Artifact error: from_estimator() got an unexpected keyword argument 'y_true'
```

File: mlflow/sklearn/utils.py [hunk 11]
```diff
@@ -310,19 +312,24 @@ def plot_confusion_matrix(*args, **kwargs):
...
             }
         ):
-            return sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            return (
+                sklearn.metrics.ConfusionMatrixDisplay.from_estimator(*args, **kwargs)
+                if is_plot_function_deprecated
+                else sklearn.metrics.plot_confusion_matrix(*args, **kwargs)
+            )
 
+    target_label_arg_name = "y" if is_plot_function_deprecated else "y_true"
```
Comment (harupy) [inline mlflow/sklearn/utils.py:321]: @shaikmoeed `sklearn.metrics.ConfusionMatrixDisplay.from_estimator` uses `y` for target labels but `sklearn.metrics.plot_confusion_matrix` uses `y_true`.

Comment (harupy) [reply mlflow/sklearn/utils.py]: This should fix the `Cross version tests / test (sklearn / 1.0.2 / autologging)` failure.

Comment (shaikmoeed) [reply mlflow/sklearn/utils.py]: Thanks for making changes and explaining them. I got a better picture/learning to contribute next time. I hope, all the test cases run successfully :)

**Comment by harupy** [comment]
All the sklearn tests finally passed!

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/runs/6037528438?check_suite_focus=true#step:12:19

image

All green (= no warnigs) 🎉

**Review by harupy — APPROVED** [review]
LGTM!
